crates/ty_python_semantic/src/types.rs

```diff
@@ -1452,6 +1568,112 @@ impl<'db> Type<'db> {
         }
     }
 
+    #[must_use]
+    pub(crate) fn recursive_type_normalized(
+        self,
+        db: &'db dyn Db,
+        cycle_heads: &salsa::CycleHeads,
+    ) -> Self {
+        cycle_heads.ids().fold(self, |ty, id| {
+            let visitor = RecursiveTypeNormalizedVisitor::new(Type::divergent(id));
+            ty.recursive_type_normalized_impl(db, &visitor)
+        })
+    }
+
+    fn recursive_type_normalized_impl(
+        self,
+        db: &'db dyn Db,
+        visitor: &RecursiveTypeNormalizedVisitor<'db>,
+    ) -> Self {
+        if visitor.level() == 0 && self == visitor.div {
+            // int | Divergent = int | (int | (int | ...)) = int
+            return Type::Never;
```

Comment:
I don't think I understand this comment, specifically how it is related to union handling, given that `level == 0`